src/EditorFeatures/Core/Structure/AbstractStructureTaggerProvider.cs

```diff
@@ -67,37 +67,146 @@ protected override bool ComputeInitialTagsSynchronously(ITextBuffer subjectBuffe
                 return false;
 
             // If we're a metadata-as-source doc, we need to compute the initial set of tags synchronously
             // so that we can collapse all the .IsImplementation tags to keep the UI clean and condensed.
-            var isMetadataAsSource = openDocument.Project.Solution.Workspace.Kind == WorkspaceKind.MetadataAsSource;
-            if (isMetadataAsSource)
+            if (openDocument.Project.Solution.Workspace.Kind == WorkspaceKind.MetadataAsSource &&
+                GlobalOptions.GetOption(BlockStructureOptionsStorage.CollapseMetadataImplementationsWhenFirstOpened, openDocument.Project.Language))
+            {
                 return true;
+            }
+
+            // If the user wants to collapse usings or #regions then we need to compute
+            // synchronously, but only if there are usings or #regions in the file. To
+            // save some work, we'll look for both in a single pass.
+            var collapseRegions = GlobalOptions.GetOption(BlockStructureOptionsStorage.CollapseRegionsWhenFirstOpened, openDocument.Project.Language);
+            var collapseUsings = GlobalOptions.GetOption(BlockStructureOptionsStorage.CollapseImportsWhenFirstOpened, openDocument.Project.Language);
+
+            if (!collapseRegions && !collapseUsings)
+            {
+                return false;
+            }
 
-            // If we contain any #region sections, we want to collapse those automatically on open the first
-            // time a doc is ever opened.  So we need to compute the initial tags synchronously in order to
-            // do that.
-            if (ContainsRegionTag(subjectBuffer.CurrentSnapshot))
+            if (ContainsRegionOrUsing(subjectBuffer.CurrentSnapshot, collapseRegions, collapseUsings, openDocument.Project.Language))
+            {
                 return true;
+            }
 
             return false;
+        }
 
-            static bool ContainsRegionTag(ITextSnapshot textSnapshot)
+        // Internal for testing. Can't use TestAccessor because the base class already defines one
+        internal static bool ContainsRegionOrUsing(ITextSnapshot textSnapshot, bool collapseRegions, bool collapseUsings, string language)
+        {
+            foreach (var line in textSnapshot.Lines)
             {
-                foreach (var line in textSnapshot.Lines)
+                if (collapseRegions && StartsWithRegionTag(line))
+                {
+                    return true;
+                }
+                else if (collapseUsings && IsUsingDeclarationOrExtern(line, language))
                 {
-                    if (StartsWithRegionTag(line))
-                        return true;
+                    return true;
                 }
+            }
+
+            return false;
+
+            static bool StartsWithRegionTag(ITextSnapshotLine line)
+            {
+                if (line.Length < 7)
```

Comment:
```suggestion
                if (line.Length < "#region".Length)
```

also, introduce constnat for this.